test/websocket.misc.js

```diff
@@ -0,0 +1,37 @@
+var assert = require('assert');
+var utils = require('./helpers/test.utils');
+var Web3 = utils.getWeb3();
+
+describe('Websocket etc & errors', function () {
+    let web3;
+
+    beforeEach(() => web3 = new Web3('ws://localhost:' + 8777));
+
+    // This test's error is fired by the request queue checker in the onClose handler
+    it('errors when there is no connection', async function(){
```

Comment:
The ``connection not open`` error doesn't get triggered from the ``onClose`` listener this error does get thrown in the ``send`` method of the provider.

Reply:
I've reverted to the old logic in the latest commit, while keeping the test expectation so you can see this is not the case [in travis here](https://travis-ci.org/ethereum/web3.js/jobs/635338577#L3658-L3682).

In the current 1.x branch, close doesn't throw any errors itself but there is a change in the readyState which causes the request to hit the logic in send. That's no longer the case on the provider-improvements branch.

Reply:
The described behaviour from you only happens if the underlying ``WebSocket`` connection does have the ``readyState`` set to ``0`` when calling the ``send`` method of the ``WebsocketProvider``. If the provider does have the ``readyState`` sett to ``2`` or ``3`` will it throw the ``connection not open`` error but also only if the provider didn't start to reconnect. Which means it is the same logic as before only with the ``reconnecting`` case added. (see: https://github.com/ethereum/web3.js/blob/1.x/packages/web3-providers-ws/src/index.js#L252)

Reply:
Do you remain concerned that the wrong error will be fired with the proposed change to the request clearance? Or do you think the change makes sense?

Reply:
hmm.. we could keep the error in the close listener as it is and extend the conditions in the ``WebsocketProvider.send`` to throw the ``connection not open`` error when the ``readyState`` is set to ``CLOSING``. If I'm right is it with this addition the same "error logic" as in the currently released version. Let me check that quickly.

Reply:
I was sadly wrong the current conditions do already handling the ``CLOSING`` state correctly.

Reply:
The current logic would be to throw the ``ConnectionNotOpenError`` for all items in the ``requestQueue`` as you already have done. We could there extend the error object with additional properties as for example the ``code`` and ``reason`` from the close event. The ``responseQueue`` should throw the ``InvalidConnection`` error and also there could we add the ``code`` and ``reason`` property from the close event to the error object. 

WDYT?

Reply:
@nivida Yes this sounds reasonable to me - are you open to merging this as a correct description of 1.x's behavior for the most common error states and making those changes in a separate set of commits (maybe even a PR directed at #3190)?

Reply:
@cgewecke Yes, sounds good to me! 👌